clippy_lints/src/useless_conversion.rs

```diff
@@ -102,6 +103,17 @@ fn into_iter_deep_call<'hir>(cx: &LateContext<'_>, mut expr: &'hir Expr<'hir>) -
     (expr, depth)
 }
 
+/// Checks if the given `expr` is an argument of a macro invocation.
+/// This is a slow-ish operation, so consider calling this late
+/// to avoid slowing down the lint in the happy path when not emitting a warning
+fn is_macro_argument(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```

Comment:
Could we avoid that by keeping an `Option<Span>` in our lint to be set when we encounter an expression from a macro? That way this check would become a simple `.is_some()`.

Reply:
I'm not sure. How/When does it get reset to `None`? If we just set it to `Some(span)` once we see an expression from an expansion, it would forever think we are in a macro, no? Or do you mean we should also then check if the span in the lint struct fully encloses the span currently being looked at, or something like that?

Reply:
Yeah, we can have `check_expr` set the span to `Some(span)` and then remove it if it matches in `check_expr_post`.

Reply:
Hmm, but wouldn't this at least also need a stack of spans, to handle nested macro expansions? Alternatively, could we just use a `u32` that we increment/decrement for entering or leaving a span from an expansion and check if `== 0` when about to lint?